### PR TITLE
fix: ignore empty labels during label merge and templating

### PR DESCRIPTION
**Pull request by ivankatliarchuk**
## Changes

When any string in `label` or `andLabel` is empty, Github API failes with `Response code 422 (Unprocessable Entity)`. The fix should improve label API resiliency, as it possible, that during runtime some labels may not be set, or empty.

The code 

```js
return [...new Set([...labels, ...addLabels])]
      template.compile(label, config)
     .filter((el) => is.nonEmptyString(el) && !is.emptyStringOrWhitespace(el))
    .map((label) => template.compile(label, config))
    .filter((el) => is.nonEmptyString(el) && !is.emptyStringOrWhitespace(el));
  }
}
```

The first filter makes sure that input labels are not empty, not nuls and etc. In order to provide flexible configuration, Renovate supports using "templates", and when the value is not exist, not set or does not apply for a use case, `handlebars` return an empty string. Github API does not let to set empty label.

## Context

_Describe the bug_

Any of `depType` , `datasource` or `updateType` empty at execution time.
```js
module.exports = { 
  "labels": ["renovate"] ,
  "packageRules": [
    {
      "matchUpdateTypes": ["major", "minor", "patch", "pin", "digest"],
      "addLabels": ["{{depType}}", "{{datasource}}", "{{updateType}}"]
    },
  ]
}
```

or

Environment variable `OPTIONAL_TEAM_LABEL` is optional for team B but mandatory for team A 

```js
module.exports = { 
  "labels": [process.env.OPTIONAL_TEAM_LABEL] ,
}
```

_Relevant debug logs_

```
DEBUG: Adding labels 'renovate, , helm, major' to #7 (repository=ik-workshop/workshop-renovate, branch=renovate/sandbox/ex6-datadog-2.x)
DEBUG: Received invalid response - aborting (repository=ik-workshop/workshop-renovate, branch=renovate/sandbox/ex6-datadog-2.x)
       "err": {
         "name": "HTTPError",
         "code": "ERR_NON_2XX_3XX_RESPONSE",
         "timings": {
           "start": 1644788915104,
           "socket": 1644788915105,
           "lookup": 1644788915107,
           "connect": 1644788915138,
           "secureConnect": 1644788915169,
           "upload": 1644788915169,
           "response": 1644788915473,
           "end": 1644788915473,
           "phases": {
             "wait": 1,
             "dns": 2,
             "tcp": 31,
             "tls": 31,
             "request": 0,
             "firstByte": 304,
             "download": 0,
             "total": 369
           }
         },
         "message": "Response code 422 (Unprocessable Entity)",
         "stack": "HTTPError: Response code 422 (Unprocessable Entity)\n    at Request.<anonymous> (/usr/src/app/node_modules/got/dist/source/as-promise/index.js:117:42)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)",
         "options": {
           "headers": {
             "user-agent": "RenovateBot/31.72.2 (https://github.com/renovatebot/renovate)",
             "accept": "application/vnd.github.v3+json",
             "authorization": "***********",
             "content-type": "application/json",
             "content-length": "30",
             "accept-encoding": "gzip, deflate, br"
           },
           "url": "https://api.github.com/repos/ik-workshop/workshop-renovate/issues/7/labels",
           "hostType": "github",
           "username": "",
           "password": "",
           "method": "POST",
           "http2": false
         },
         "response": {
           "statusCode": 422,
           "statusMessage": "Unprocessable Entity",
           "body": {
             "message": "Validation Failed",
             "errors": [
               {
                 "value": "major",
                 "resource": "Label",
                 "field": "name",
                 "code": "invalid"
               }
             ],
             "documentation_url": "https://docs.github.com/rest/reference/issues#add-labels-to-an-issue"
           },
           "headers": {
             "server": "GitHub.com",
             "date": "Sun, 13 Feb 2022 21:48:35 GMT",
             "content-type": "application/json; charset=utf-8",
             "content-length": "202",
             "x-oauth-scopes": "repo, workflow",
             "x-accepted-oauth-scopes": "",
             "github-authentication-token-expiration": "2022-03-12 09:39:59 UTC",
             "x-github-media-type": "github.v3; format=json",
             "x-ratelimit-limit": "5000",
             "x-ratelimit-remaining": "4936",
             "x-ratelimit-reset": "1644791050",
             "x-ratelimit-used": "64",
             "x-ratelimit-resource": "core",
             "access-control-expose-headers": "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset",
             "access-control-allow-origin": "*",
             "strict-transport-security": "max-age=31536000; includeSubdomains; preload",
             "x-frame-options": "deny",
             "x-content-type-options": "nosniff",
             "x-xss-protection": "0",
             "referrer-policy": "origin-when-cross-origin, strict-origin-when-cross-origin",
             "content-security-policy": "default-src 'none'",
             "vary": "Accept-Encoding, Accept, X-Requested-With",
             "x-github-request-id": "D677:DFBB:20D48F1:216B744:62097CB3",
             "connection": "close"
           },
           "httpVersion": "1.1"
         }
       }
DEBUG: Pull request creation error (repository=ik-workshop/workshop-renovate, branch=renovate/sandbox/ex6-datadog-2.x)
       "err": {
         "message": "repository-changed",
         "stack": "Error: repository-changed\n    at handleGotError (/usr/src/app/node_modules/renovate/lib/util/http/github.ts:132:14)\n    at GithubHttp.request (/usr/src/app/node_modules/renovate/lib/util/http/github.ts:329:13)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at GithubHttp.requestJson (/usr/src/app/node_modules/renovate/lib/util/http/index.ts:238:17)\n    at addLabels (/usr/src/app/node_modules/renovate/lib/platform/github/index.ts:1259:5)\n    at Proxy.createPr (/usr/src/app/node_modules/renovate/lib/platform/github/index.ts:1518:3)\n    at ensurePr (/usr/src/app/node_modules/renovate/lib/workers/pr/index.ts:441:14)\n    at processBranch (/usr/src/app/node_modules/renovate/lib/workers/branch/index.ts:620:33)\n    at writeUpdates (/usr/src/app/node_modules/renovate/lib/workers/repository/process/write.ts:38:17)\n    at update (/usr/src/app/node_modules/renovate/lib/workers/repository/process/extract-update.ts:129:11)\n    at Object.renovateRepository (/usr/src/app/node_modules/renovate/lib/workers/repository/index.ts:46:17)\n    at Object.start (/usr/src/app/node_modules/renovate/lib/workers/global/index.ts:126:7)\n    at /usr/src/app/node_modules/renovate/lib/renovate.ts:16:22"
       }
```

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — CHANGES_REQUESTED** [review]
This change affects more than just "don't set empty labels" and could have unforeseen reason implications. Please either fix the narrow scope or else start an issue to discuss wider requirements.

I also recommend you simplify to use is.truthy and is.nonEmptyString

File: lib/util/template/index.ts [hunk 1]
```diff
@@ -161,9 +161,9 @@ function getFilteredObject(input: CompileInput): FilteredObject {
   for (const field of allAllowed) {
     const value = obj[field];
     if (is.array(value)) {
-      res[field] = value.map((element) =>
-        getFilteredObject(element as CompileInput)
-      );
+      res[field] = value
+        .filter((el) => el !== null)
```
Comment (viceice) [inline lib/util/template/index.ts:165]:
```suggestion
        .filter(is.object)
```

File: lib/workers/pr/index.ts [hunk 2]
```diff
@@ -56,9 +56,10 @@ function prepareAssigneesReviewers(
 export function prepareLabels(config: RenovateConfig): string[] {
   const labels = config.labels ?? [];
   const addLabels = config.addLabels ?? [];
-  return [...new Set([...labels, ...addLabels])].map((label) =>
-    template.compile(label, config)
-  );
+  return [...new Set([...labels, ...addLabels])]
+    .filter((el) => el !== null && el !== undefined && el.trim() !== '')
```
Comment (viceice) [inline lib/workers/pr/index.ts:60]:
```suggestion
    .filter(is.nonEmptyString)
```

File: lib/workers/pr/index.ts [hunk 3]
```diff
@@ -56,9 +56,10 @@ function prepareAssigneesReviewers(
...
+    .map((label) => template.compile(label, config))
+    .filter((el) => el !== null && el.trim() !== '');
```
Comment (viceice) [inline lib/workers/pr/index.ts:62]:
```suggestion
    .filter(is.nonEmptyString);
```

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/workers/pr/index.ts (hunk 2 above)
Comment (ivankatliarchuk) [reply lib/workers/pr/index.ts:60]: Tests are failing. In this case, `is.nonEmptyString` does not cover a case when string contains multiple white space character.
It still require `trim` e.g. `.filter((el) => is.nonEmptyString(el) && el.trim() !== '')`

Comment (ivankatliarchuk) [reply lib/workers/pr/index.ts]: Thanks for help. Ive updated the logic with `.filter((el) => is.nonEmptyString(el) && !is.emptyStringOrWhitespace(el))` and created a proposal to add `nonEmptyStringOrWhitespace` https://github.com/sindresorhus/is/issues/158

File: lib/workers/pr/index.ts (hunk 3 above)
Comment (ivankatliarchuk) [reply lib/workers/pr/index.ts]: updated with `.filter((el) => is.nonEmptyString(el) && !is.emptyStringOrWhitespace(el))`

**Comment by ivankatliarchuk** [comment]
Thanks for the suggestions. Can I have please an idea of how to reduce the scope for the bug described?
Created an issue https://github.com/renovatebot/renovate/issues/14328

**Comment by HonkingGoose** [comment]
Hi @ivankatliarchuk! 👋 

Thank you for your contribution!

You can mark a PR as a "draft" from within GitHub. You don't need to put `(draft)` in the PR title. 😉 

Please read the [GitHub docs, Changing the stage of a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request), and then mark the PR as draft within GitHub, and remove the `(draft)` text from the PR title.

**Review by rarkins — DISMISSED** [review]

**Comment by HonkingGoose** [comment]
@rarkins don't you want to edit the PR title first, before you merge?

File: lib/util/template/index.ts [hunk 4]
```diff
@@ -161,9 +161,9 @@ function getFilteredObject(input: CompileInput): FilteredObject {
   for (const field of allAllowed) {
     const value = obj[field];
     if (is.array(value)) {
-      res[field] = value.map((element) =>
-        getFilteredObject(element as CompileInput)
-      );
+      res[field] = value
+        .filter(is.object)
```
Comment (viceice) [inline lib/util/template/index.ts:165]:
```suggestion
        .filter(is.plainObject)
```
Sorry, seen this function some lines below.

File: lib/workers/pr/index.ts [hunk 5]
```diff
@@ -56,9 +57,10 @@ function prepareAssigneesReviewers(
 export function prepareLabels(config: RenovateConfig): string[] {
   const labels = config.labels ?? [];
   const addLabels = config.addLabels ?? [];
-  return [...new Set([...labels, ...addLabels])].map((label) =>
-    template.compile(label, config)
-  );
+  return [...new Set([...labels, ...addLabels])]
+    .filter((el) => is.nonEmptyString(el) && !is.emptyStringOrWhitespace(el))
```
Comment (viceice) [inline lib/workers/pr/index.ts:61]: Can you extract this to a small helper function and use it in both places? Would make the code more readable.

**Review by viceice — COMMENTED** [review]
Otherwise LGTM

**Comment by ivankatliarchuk** [comment]
Hi, @HonkingGoose.  Apologies spent too much time working on GitLab. The Draft button there is a bit more visible. I found it btw -;)

File: lib/workers/pr/index.ts (hunk 5 above)
Comment (ivankatliarchuk) [reply lib/workers/pr/index.ts]: added `nonEmptyStringAndNotWhitespace` function

File: lib/workers/pr/index.ts [hunk 6]
```diff
@@ -29,6 +30,10 @@ function noLeadingAtSymbol(input: string): string {
   return input.length && input.startsWith('@') ? input.slice(1) : input;
 }
 
+function nonEmptyStringAndNotWhitespace(input: string): boolean {
```
Comment (viceice) [inline lib/workers/pr/index.ts:33]: Sorry, but please move it to `lib/util/string`. Seems useful for other places to me.

Comment (ivankatliarchuk) [reply lib/workers/pr/index.ts]: Just did. Hard to say. `is` library seems the right place for it https://github.com/sindresorhus/is/issues/158

Comment (viceice) [reply lib/workers/pr/index.ts]: Sure, but we don't know when it will be available  🤷‍♂️🤗

**Review by viceice — APPROVED** [review]

**Pull request merged**